<a href="https://colab.research.google.com/github/Saif-Khan435/Play-Store-App-Review-Analysis/blob/main/Mohammed_Saif_Khan_M1W6_Play_Store_App_Review_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problem Statement :**

## <b> The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market. </b>

## <b> Each app (row) has values for catergory, rating, size, and more. Another dataset contains customer reviews of the android apps.</b>

## <b> Explore and analyze the data to discover key factors responsible for app engagement and success. </b>

# **Introduction :**

We are provided with two datasets, one contains all the details of the app, and another one contains the user reviews of the respective app. By analysing these two datasets, we must come up with the key factors that are responsible for app engagement and success.

Steps to follow to find key factors responsible for app engagement and success:-
* Loading the Libraries
* Loading the Dataset
* Exploring the Dataset
* Looking for and Handling NaN/Null/Missing Values
* Cleaning the DataFrame
* Data Visualization
* Analysis Summary
* Conclusion


# **Loading the Libraries :**

In [3]:
# Loading the Libraries required for Exploratory Data Analysis

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn

Numpy and Pandas Libraries are essential for Data Analysis. Numpy is used for working on N-Dimensional Ararys and performing fast Numerical Calculations. Pandas is used for Data Manipulation and Analysis.

We require libraries like Matplotlib and Seaborn for creating Data Visualization. %matplotlib inline makes your plot outputs appear and be stored within the notebook itself.

In [4]:
import warnings
warnings.filterwarnings('ignore') # The warnings raised during runtime are not displayed

# **Loading the Datasets :**

In [5]:
# Mounting the Drive

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [87]:
# Loading the Datasets using Pandas

apps_df = pd.read_csv('/content/drive/MyDrive/Alma Better Pinnacle Pro/Module 1 - Capstone Projects/Play Store App Review Analysis/Copy of Play Store Data.csv')
reviews_df = pd.read_csv('/content/drive/MyDrive/Alma Better Pinnacle Pro/Module 1 - Capstone Projects/Play Store App Review Analysis/Copy of User Reviews.csv')

The two datasets have been loaded (read) by using read_csv function from pandas library now we can begin the exploration of the datasets.

# **Exploring the Datasets :**

## **Exploring the Apps Dataframe :**

In [10]:
apps_df.head(10) # Displays the first 10 rows

App  ...   Android Ver
0     Photo Editor & Candy Camera & Grid & ScrapBook  ...  4.0.3 and up
1                                Coloring book moana  ...  4.0.3 and up
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ...  4.0.3 and up
3                              Sketch - Draw & Paint  ...    4.2 and up
4              Pixel Draw - Number Art Coloring Book  ...    4.4 and up
5                         Paper flowers instructions  ...    2.3 and up
6            Smoke Effect Photo Maker - Smoke Editor  ...  4.0.3 and up
7                                   Infinite Painter  ...    4.2 and up
8                               Garden Coloring Book  ...    3.0 and up
9                      Kids Paint Free - Drawing Fun  ...  4.0.3 and up

[10 rows x 13 columns]

In [11]:
apps_df.tail(10) # Displays the last 10 rows

App  ...         Android Ver
10831                       payermonstationnement.fr  ...          4.0 and up
10832                                       FR Tides  ...          2.1 and up
10833                                    Chemin (fr)  ...          2.2 and up
10834                                  FR Calculator  ...          4.1 and up
10835                                       FR Forms  ...          4.0 and up
10836                               Sya9a Maroc - FR  ...          4.1 and up
10837               Fr. Mike Schmitz Audio Teachings  ...          4.1 and up
10838                         Parkinson Exercices FR  ...          2.2 and up
10839                  The SCP Foundation DB fr nn5n  ...  Varies with device
10840  iHoroscope - 2018 Daily Horoscope & Astrology  ...  Varies with device

[10 rows x 13 columns]

In [12]:
apps_df.shape # Returns the shape of the dataframe

(10841, 13)

In [13]:
apps_df.columns # Returns the names of the columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [14]:
apps_df.info() # Returns the information about the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [16]:
apps_df.describe(include='all') # Returns the descriptive statistics of the dataframe

App Category  ...         Current Ver Android Ver
count    10841    10841  ...               10833       10838
unique    9660       34  ...                2832          33
top     ROBLOX   FAMILY  ...  Varies with device  4.1 and up
freq         9     1972  ...                1459        2451
mean       NaN      NaN  ...                 NaN         NaN
std        NaN      NaN  ...                 NaN         NaN
min        NaN      NaN  ...                 NaN         NaN
25%        NaN      NaN  ...                 NaN         NaN
50%        NaN      NaN  ...                 NaN         NaN
75%        NaN      NaN  ...                 NaN         NaN
max        NaN      NaN  ...                 NaN         NaN

[11 rows x 13 columns]

**Let us first define what information the columns contain based on our inspection.**

* `App`: Contains the name of the app with a short description (optional).
* `Category`: It gives the category to the app.
* `Rating`: It contains the average rating the respective app received from its users.
* `Reviews`: It contains the number of users that have dropped a review for the respective app.
* `Size`: It contains the disk space required to install the respective app.
* `Installs`: It gives the rounded figure of number of times the respective app was downloaded.
* `Type`: It states whether an app is free to use or paid.
* `Price`: It gives the price payable to install the app. For free type apps, the price is zero.
* `Content rating`: It states whether or not an app is suitable for all age groups or not.
* `Genres`: It gives the genre(s) to which the respective app belongs.
* `Last updated`: It gives the day in which the latest update was released.
* `Current Ver`: It gives the current version of the respective app.
* `Android Ver`: It gives the android version of the respective app.

## **Exploring the Reviews Dataframe :**

In [18]:
reviews_df.head(10) # Displays the first 10 rows

App  ... Sentiment_Subjectivity
0  10 Best Foods for You  ...               0.533333
1  10 Best Foods for You  ...               0.288462
2  10 Best Foods for You  ...                    NaN
3  10 Best Foods for You  ...               0.875000
4  10 Best Foods for You  ...               0.300000
5  10 Best Foods for You  ...               0.300000
6  10 Best Foods for You  ...               0.900000
7  10 Best Foods for You  ...                    NaN
8  10 Best Foods for You  ...               0.000000
9  10 Best Foods for You  ...               0.000000

[10 rows x 5 columns]

In [19]:
reviews_df.tail(10) # Displays the last 10 rows

App  ... Sentiment_Subjectivity
64285  Houzz Interior Design Ideas  ...                    NaN
64286  Houzz Interior Design Ideas  ...                    NaN
64287  Houzz Interior Design Ideas  ...                    NaN
64288  Houzz Interior Design Ideas  ...                    NaN
64289  Houzz Interior Design Ideas  ...                    NaN
64290  Houzz Interior Design Ideas  ...                    NaN
64291  Houzz Interior Design Ideas  ...                    NaN
64292  Houzz Interior Design Ideas  ...                    NaN
64293  Houzz Interior Design Ideas  ...                    NaN
64294  Houzz Interior Design Ideas  ...                    NaN

[10 rows x 5 columns]

In [20]:
reviews_df.shape # Returns the shape of the dataframe

(64295, 5)

In [21]:
reviews_df.columns # Returns the names of the columns

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')

In [22]:
reviews_df.info() # Returns the information about the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [23]:
reviews_df.describe(include='all') # Returns the descriptive statistics of the dataframe

App  ... Sentiment_Subjectivity
count        64295  ...           37432.000000
unique        1074  ...                    NaN
top     Bowmasters  ...                    NaN
freq           320  ...                    NaN
mean           NaN  ...               0.492704
std            NaN  ...               0.259949
min            NaN  ...               0.000000
25%            NaN  ...               0.357143
50%            NaN  ...               0.514286
75%            NaN  ...               0.650000
max            NaN  ...               1.000000

[11 rows x 5 columns]

**Let us first define what information the columns contain based on our inspection.**

* `App`: Contains the name of the app with a short description (optional).
* `Translated_Review`: It contains the English translation of the review dropped by the user of the app.
* `Sentiment`: It gives the attitude/emotion of the writer. It can be ‘Positive’, ‘Negative’, or ‘Neutral’.
* `Sentiment_Polarity`: It gives the polarity of the review. Its range is [-1,1], where 1 means ‘Positive statement’ and -1 means a ‘Negative statement’.
* `Sentiment_Subjectivity`: This value gives how close a reviewers opinion is to the opinion of the general public. Its range is [0,1]. Higher the subjectivity, closer is the reviewers opinion to the opinion of the general public, and lower subjectivity indicates the review is more of a factual information.

# **Looking for and Handling NaN/Null/Missing Values**

## **Looking for and Handling NaN/Null/Missing Values in Apps Dataframe :**

In [88]:
# Find the total number of NaN values in each column

apps_df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

Let us try to handle these NaN values one column at a time.

Lets find the percentage of of rows that contain atleast one NaN value.

In [65]:
# Percentage of rows containing atleast one NaN value

total_nan = sum([True for id,row in apps_df.iterrows() if any(row.isnull())])

perc_total_nan = round(total_nan/10841*100, 2)

print(f'Hence, there are a total of {total_nan}({perc_total_nan}%) rows which contain atleast one NaN value.')

Hence, there are a total of 1481(13.66%) rows which contain atleast one NaN value.


Since approximately 13.66% of all the rows in the dataset contains atleast one NaN value, it is not practcal to eliminate all the rows containing NaN values. We need to handle this problem in another way.

Lets first deal with the columns that have the least number of NaN values. By going through the NaN values, we must come up with a way to replace them with non NaN values or we need to come up with a reason for having NaN.

**1. `Type`: There is only 1 NaN values in this column.**

In [89]:
# The row containing NaN values in the Type column

apps_df[apps_df.Type.isnull()]

App  ...         Android Ver
9148  Command & Conquer: Rivals  ...  Varies with device

[1 rows x 13 columns]

In [90]:
# Finding the different values the 'Type' column takes

apps_df.Type.value_counts()

Free    10039
Paid      800
0           1
Name: Type, dtype: int64

The `Type` column contains only two entries, namely, `Free` and `Paid`. Also, if the app is of type-`paid`, the price of that app will be printed in the corresponding `Price` column, else, it will show as '0'. In this case, the price for the respective app is printed as '0', which means the app is of type-`free`. Hence we can replace this NaN value with `Free`.

In [91]:
# Replacing the NaN value in 'Type' column with 'Free'

apps_df.Type = apps_df.Type.replace(np.NaN,'Free')

In [92]:
# Total number of NaN values in each column

apps_df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

We were successfully able to handle the NaN values in the `Type` column.

**1. `Content Rating`: There is only 1 NaN values in this column.**

In [96]:
# The row containing NaN values in the Content Rating column

apps_df[apps_df['Content Rating'].isnull()]

App Category  ...  Current Ver Android Ver
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9  ...   4.0 and up         NaN

[1 rows x 13 columns]

As we can see the `Category` contains the value '1.9', which does not make any sense, the `Rating` is '19' which cannot be true, since the play store app reviews can be in between 1 and 5 only, and so on. So it is of our best interest to delete this row altogather so as to eliminate any errors as we proceed with the analysis.

In [84]:
# Drop the row with the NaN value i.e. row index = 10472

apps_df.dropna(subset=['Content Rating'],inplace=True)

In [85]:
# Total number of NaN values in each column

apps_df.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

We were successfully able to handle the NaN values in the `Content Rating` column.

**3. `Android Ver`: There are only 2 NaN values in this column.**

**4. `Current Ver`: There are 8 NaN values in this column.**

**5. `Rating`: There are 1470 NaN values in this column.**

## **Looking for and Handling NaN/Null/Missing Values in Reviews Dataframe :**

In [41]:
# Find the total number of NaN values in each column

reviews_df.isnull().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

# **Cleaning the DataFrames :**

## **Cleaning the Apps Dataframe :**

## **Cleaning the Reviews Dataframe :**